### PR TITLE
input: add an input-evdev device

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Add a device driver to read events from a Linux evdev device node and inject them back as Zephyr input events.

### CODE REVIEW

**Comment by fabiobaltieri** [comment]
Hey @aescolar do I need to do a top/bottom split for this one?

**Comment by aescolar** [comment]
> Hey @aescolar do I need to do a top/bottom split for this one?

Yes please :)

File: dts/bindings/input/input-evdev.yaml
Comment (aescolar) [inline dts/bindings/input/input-evdev.yaml:19]: I guess the user will need to sudo to be able to read an input device? (anyhow, better described in docs :) )

**Review by aescolar — CHANGES_REQUESTED** [review]
Thanks @fabiobaltieri 

A few comments, most of them quite trivial.

May I ask for :
* A paragraph in https://docs.zephyrproject.org/latest/boards/posix/native_sim/doc/index.html#peripherals 
* An entry in https://docs.zephyrproject.org/latest/boards/posix/native_sim/doc/index.html#peripherals-and-backends-c-library-compatibility next to the SDL input one

File: dts/bindings/input/input-evdev.yaml
Comment (fabiobaltieri) [reply dts/bindings/input/input-evdev.yaml]: yes, done

File: dts/bindings/input/input-evdev.yaml
Comment (fabiobaltieri) [reply dts/bindings/input/input-evdev.yaml:10]: Don't think links works great in bindings, like they don't get referenced in the generated doc etc, I'd keep it on both

File: drivers/input/input_evdev.c
Comment (fabiobaltieri) [reply drivers/input/input_evdev.c]: brilliant! thanks, changed

File: drivers/input/input_evdev.c
Comment (fabiobaltieri) [reply drivers/input/input_evdev.c]: oh nice, done

File: drivers/input/input_evdev.c
Comment (fabiobaltieri) [reply drivers/input/input_evdev.c]: fixed

File: drivers/input/input_evdev.c
Comment (fabiobaltieri) [reply drivers/input/input_evdev.c]: fixed, kept the "one", this driver really wants one if it's in the build

File: drivers/input/input_evdev_bottom.h
Comment (fabiobaltieri) [reply drivers/input/input_evdev_bottom.h:6]: added the guards, don't see much of a point having the c++ one, can't see how this could be included in a c++ build

File: drivers/input/input_evdev_bottom.c
Comment (fabiobaltieri) [reply drivers/input/input_evdev_bottom.c:21]: Oh very good, thanks for the clear example, I thought that this was going to run in its own pthread or something. Alright good to know, applied my interpretation of your suggested change, lmk what you think

File: drivers/input/input_evdev.c
Comment (fabiobaltieri) [reply drivers/input/input_evdev.c]: oh ok, I added a define and set to 10ms, don't see why one would change so I'd avoid the extra option, btw I copied this from `can_native_linux.c`... maybe there should be a common `nsi_...` function for this?

File: dts/bindings/input/input-evdev.yaml
Comment (fabiobaltieri) [reply dts/bindings/input/input-evdev.yaml:19]: Ahh... maybe? :-) I don't, guess I added my user to the input group on all my systems. Honestly I'd rather leave this up to the user to figure out rather than adding a Linux system howto in every driver

File: boards/posix/native_sim/doc/index.rst
Comment (aescolar) [inline boards/posix/native_sim/doc/index.rst:452]: Nit: Maybe "flag" => "command line option"? (just to be more specific)

**Comment by fabiobaltieri** [comment]
Thanks @aescolar, reworked quite few things, you may want to review from scratch. Two open questions:
- should `nsi_host_open` be used for opening the device? Can't see any example
- should this be renamed to `linux-evdev`? undecided

Thanks for the good review

File: drivers/input/input_evdev_bottom.c
Comment (aescolar) [reply drivers/input/input_evdev_bottom.c:21]: Just for info: It is its own thread, but native_sim only has a single "core" emulated MCU, so if you hung that thread in a call to the Linux OS, you hung everything.

**Review by aescolar — COMMENTED** [review]
Looking good, just a trivial thing

**Comment by aescolar** [comment]
> should nsi_host_open be used for opening the device? Can't see any example

You would not have the flags from the top context, as you don't have the includes. So nope.

>  should this be renamed to linux-evdev? undecided

I would lean towards yes :)

File: drivers/input/input_evdev.c
Comment (aescolar) [reply drivers/input/input_evdev.c]: @henrikbrixandersen FYI :), this comment applies also to the wait in the can_native_linux driver loop :)

>  maybe there should be a common nsi_... function for this?

I guess what you want to do in the receiving thread can be quite dependent of the type of driver.

File: boards/posix/native_sim/native_sim.dts
Comment (fabiobaltieri) [inline boards/posix/native_sim/native_sim.dts:20]: woops, sorry this was not mean to be here

File: boards/posix/native_sim/doc/index.rst
Comment (fabiobaltieri) [reply boards/posix/native_sim/doc/index.rst]: ok fixed

**Comment by fabiobaltieri** [comment]
- renamed to linux-evdev
- moved the wait in the "no data" case, this allows handling bursts of data with no delay

**Review by aescolar — DISMISSED** [review]

File: dts/bindings/input/linux-evdev.yaml
Comment (henrikbrixandersen) [inline dts/bindings/input/linux-evdev.yaml:21]: Should this be `zephyr,linux-evdev`?

File: drivers/input/Kconfig.evdev
Comment (henrikbrixandersen) [inline drivers/input/Kconfig.evdev:4]: Should this be `config NATIVE_LINUX_EVDEV`?

File: drivers/input/linux_evdev.c
Comment (faxe1008) [inline drivers/input/linux_evdev.c:105]: Maybe make the priority configurable?

File: drivers/input/linux_evdev.c
Comment (faxe1008) [inline drivers/input/linux_evdev.c:22]: Maybe make this one a kconfig option?

**Review by faxe1008 — COMMENTED** [review]
Two small nits, having some of the hardcoded values be configurable would be nice, otherwise LGTM.

File: dts/bindings/input/linux-evdev.yaml
Comment (fabiobaltieri) [reply dts/bindings/input/linux-evdev.yaml:21]: uhm good point, should probably be `zephyr,native-linux-evdev` actually

File: drivers/input/linux_evdev.c
Comment (fabiobaltieri) [reply drivers/input/linux_evdev.c]: yeah good call, reminded me I need to do it for the keyboard library as well

File: drivers/input/linux_evdev.c
Comment (fabiobaltieri) [reply drivers/input/linux_evdev.c]: I really can't imagine why one would bother changing it, but since both you and @aescolar pointed it out, I made it I a config :-)

File: drivers/input/Kconfig.evdev
Comment (fabiobaltieri) [reply drivers/input/Kconfig.evdev]: yeah good call, changed

File: dts/bindings/input/linux-evdev.yaml
Comment (fabiobaltieri) [reply dts/bindings/input/linux-evdev.yaml]: changed

**Comment by fabiobaltieri** [comment]
renamed things, added configs

**Review by gmarull — DISMISSED** [review]

**Review by faxe1008 — DISMISSED** [review]

**Review by aescolar — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Comment by carlescufi** [comment]
@fabiobaltieri please rebase

**Comment by fabiobaltieri** [comment]
rebased

**Review by aescolar — APPROVED** [review]

**Review by faxe1008 — APPROVED** [review]

**Pull request merged**
